### DISCUSSION TITLE
how to use dynamic sorting and order?

### DISCUSSION THREAD

**Discussion by kiranimeid** [porsager/postgres#655] [Q&A] [answered]
i have something like this:
sql`select * from ticket where account = ${acc.id}  order by ${orderby} ${order}`;

but when i run it, i got an error says:
'PostgresError: syntax error at or near "$2"'

I've searching on google/this repo, but i have no clue how i can achieve it. thank you.

**Comment by porsager** [comment]
I searched for "dynamic order by" on this repo, and this was the first issue in the results : https://github.com/porsager/postgres/issues/174 

I think you should find your answer there ;)

**Reply by kiranimeid** [reply]
I must be using wrong keyword then. Thank you.

**Reply by porsager** [reply, marked as answer]
No, the issue is that your orderby column is sent as a parameter. Here is your sample query which should work (notice orderby is inside `sql()`):

```js
const orderby = 'some_column'
const order = 'desc'

await sql`
  select 
    * 
  from ticket 
  where account = ${ acc.id }  
  order by ${ sql(orderby) } ${ order === 'desc' ? sql`desc` : sql`asc` }
`
```

Of course you could make the order by clause more dynamic, but that's almost always up to the specific userland implementation requirements. Imagine something like this:

```js
const id = 1
const order = {
  username: 'asc'
  created_at: 'desc'
}

await sql`
  select 
    * 
  from ticket 
  where account = ${ id }  
  order by ${
    Object.entries(order).flatMap(([column, order], i) =>
      [i ? sql`,` : sql``, sql`${ sql(column) } ${ order === 'desc' ? sql`desc` : sql`asc` }`]
    )
  }
`
```

The whole point of all this is to ensure we only have statically defined strings that compose the query. Thereby we prevent any kind of sql injection.
